#  A mystery app profile that shows huge revenue potential

##  The search for an unknown app profile with the potential for maximized profits from in-app ads

    I've gathered you here today to assist in the search of a mystery app profile with the highest potential to generate revenue from in-app advertisements.  Needless to say this is top priority and the suits from HQ want our brightest minds on this to help unravel this mystery.  The important conditions on our mystery app are that it is in english and that it is free.
    We have 3 main objectives to complete this mission:
                                                        1) Data Cleaning
                                                        2) Data Analysis
                                                        3) Building an app profile

## Data Cleaning
    We live in an imperfect world, and that means imperfect raw data.  I know it's not glamorous, but putting in    the leg work now cleaning this raw data means that we can trust the results of our analysis.  
    Our tasks are as follows:
                                            1) Removing the header row from each dataset
                                            2) Detecting and removing any rows with missing information
                                            3) Detecting and removing any duplicate rows
                                            4) Removing any non english apps
                                            5) Removing any paid apps
                                            

## 1)  
First off we must import the necessary data.  At this time we will also separate the header row from the rest of the dataset.

In [41]:
# imports appropriate module for parsing csv files
from csv import reader

#opens csv file and stores it as a list of lists
opened_file_apple = open('AppleStore.csv')
read_file_apple = reader(opened_file_apple)
apple_store_data = list(read_file_apple)
apple_store_header = apple_store_data[0] # this stores the header row seperately from the rest of the dataset
apple_store_data = apple_store_data[1:]

opened_file_google = open('googleplaystore.csv')
read_file_google = reader(opened_file_google)
google_play_data = list(read_file_google)
google_play_header = google_play_data[0] # this stores the header row seperately from the rest of the dataset
google_play_data = google_play_data[1:]

HQ (Dataquest) has also provided us with a function to explore our datasets.  This function will take the dataset and row indeces as the arguments.  There is also the optional argument for rows_and_columns, setting this argument to True will provide us with the dimensions of our datasets.  We removed the header rows so these dimensions could be as accurate as possible.

In [42]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

Okay team, with our function written let's dive into our datasets.  Nobody likes to get their boots dirty, but this is where the real work happens.

In [43]:
explore_data(apple_store_data, 0, 5, True) #displays the first 5 rows and prints the dimensions of the dataset

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']


['284035177', 'Pandora - Music & Radio', '130242560', 'USD', '0.0', '1126879', '3594', '4.0', '4.5', '8.4.1', '12+', 'Music', '37', '4', '1', '1']


Number of rows: 7197
Number of columns: 16


In [44]:
explore_data(google_play_data, 0, 5, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']


Number of rows: 10841
Number of columns: 13


And let's print our header rows as a reference for our analysis.

In [45]:
print(google_play_header)
print('\n')
print(apple_store_header)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


Excellent start! From a cursory view of our datasets we are definitely going to be interested in the ratings, genres, app name and number of reviews.

## 2)

    Now intel has shown that one of the rows in our Google play dataset has an error.  Specifically we are missing a value for one of the columns in row 10472, and this has resulted in the other columns shifting.  
    Let's investige!

In [46]:
google_play_data[10471:10474]

[['Xposed Wi-Fi-Pwd',
  'PERSONALIZATION',
  '3.5',
  '1042',
  '404k',
  '100,000+',
  'Free',
  '0',
  'Everyone',
  'Personalization',
  'August 5, 2014',
  '3.0.0',
  '4.0.3 and up'],
 ['Life Made WI-Fi Touchscreen Photo Frame',
  '1.9',
  '19',
  '3.0M',
  '1,000+',
  'Free',
  '0',
  'Everyone',
  '',
  'February 11, 2018',
  '1.0.19',
  '4.0 and up'],
 ['osmino Wi-Fi: free WiFi',
  'TOOLS',
  '4.2',
  '134203',
  '4.1M',
  '10,000,000+',
  'Free',
  '0',
  'Everyone',
  'Tools',
  'August 7, 2018',
  '6.06.14',
  '4.4 and up']]

Our intel proved correct.  We are missing the genre and the other values have shifted over.  Since we do not know the genre of this app, our best move is to delete this row and continue with our cleaning.

In [47]:
del(google_play_data[10472])

## 3)

An important part of data cleaning is to ensure that we have only unique rows, so our next step will be to create a function that detects duplicate rows and returns the name of those rows in a list.  This list can then be used to filter the duplicates our of our datasets.

In [48]:
def detect_duplicate(dataset, index):
    # creates empty lists to store the names of duplicate apps
    duplicate_apps = []
    unique_apps = []
    # loops through the dataset adding app names to the duplicate list if they are already in the unique list
    for row in dataset:
        name = row[index]
        if name in unique_apps:
            duplicate_apps.append(name)
        else:
            unique_apps.append(name)
    return duplicate_apps

In [49]:
apple_duplicate = detect_duplicate(apple_store_data, 1)
apple_duplicate

['Mannequin Challenge', 'VR Roller Coaster']

In [50]:
google_duplicate = detect_duplicate(google_play_data, 0)
len(google_duplicate)

1181

Our duplicate list for the Google play store returned so many names we decided to simply print the length of the list to keep track of how many rows we will be deleting.
Now let's look a the duplicate names from the apple store first.

In [51]:
for row in apple_store_data:
    name = row[1]
    if name == 'Mannequin Challenge' or name == 'VR Roller Coaster':
        print(row)

['1173990889', 'Mannequin Challenge', '109705216', 'USD', '0.0', '668', '87', '3.0', '3.0', '1.4', '9+', 'Games', '37', '4', '1', '1']
['952877179', 'VR Roller Coaster', '169523200', 'USD', '0.0', '107', '102', '3.5', '3.5', '2.0.0', '4+', 'Games', '37', '5', '1', '1']
['1178454060', 'Mannequin Challenge', '59572224', 'USD', '0.0', '105', '58', '4.0', '4.5', '1.0.1', '4+', 'Games', '38', '5', '1', '1']
['1089824278', 'VR Roller Coaster', '240964608', 'USD', '0.0', '67', '44', '3.5', '4.0', '0.81', '4+', 'Games', '38', '0', '1', '1']


Interesing.  Although they share the same name the ID's are different along with several other columns.  I think it is safe to say that these are all different apps and thus we will not be removing them from our dataset.

Now to tackle the google play duplicates.  There is no sense in just randomly deleting duplicates with no strategy in place, that is not how we impress the suits at HQ; we need criteria for deleting.  Since we know from our duplicate list that instagram has some copies, let's look at each copy.

In [52]:
for row in google_play_data:
    name = row[0]
    if name == 'Instagram':
        print(row)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


Here we have 4 entries for the same app, all the information is the same except for number of reviews.  This actually provides us with the perfect criterion for deleting duplicates, the row with the highest number of ratings will be the most up to date, so this will be the row that we keep.  We will do this by creating a dictionary with app names as the key and number of reviews as the value, then filter our dataset with this dictionary to keep only unique apps with the most up to date information.  To check that we did this correctly, we know that the number of rows after deleting all the duplicates should be 10840 - 1181 = 9659.

In [53]:
reviews_max = {} #empty dictionary 
# loops through the dataset, when it encounters a duplicate name it either passesd it over or updates the
#number of reviews if that number is higher than the one already in the dictionary
for row in google_play_data:
    name = row[0]
    n_reviews = float(row[3])
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
    elif name not in reviews_max:
        reviews_max[name] = n_reviews
len(reviews_max)

9659

Excellent work team!  Now we have a dictionary with only unique apps, using this we can create a dataset with only unique rows.  To do this we are going to create an empty list called android_unique to store our cleaned data.  We will also need an extra list on the off chance that two duplicates also have the same number of reviews.  A good team plans for every contingency, this is why HQ has trusted us with this important mission.

In [54]:
#creates two empty lists, one for unique apps and one just in case we have 
#two duplicates with the same number of reviews
android_unique = []
already_added = []
#loops through the dataset, only adds the rows where the app name is in the dictionary and the 
#number of reviews matches the key value
for row in google_play_data:
    name = row[0]
    n_reviews = float(row[3])
    if n_reviews == reviews_max[name] and name not in already_added:
        android_unique.append(row)
        already_added.append(name)
len(android_unique)

9659

## 4)

Now that we have taken care of the duplicates, it's time to filter out non-english apps.  To do this we are going to take advantage of the ord() function in python that gives you a unique number assigned to each character by the ASCII.  We are going to check each app name and if a character put in the ord() function returns a value greater than 127, than the character is not common in english.  The one drawback is that emoji's and certain characters that could appear in english apps have values over 127, so to help make our filter more accurate, we are only going to keep names that have 3 or less characters with values over 127.

In [55]:
#this functions loops through the individual characters in a string and determines if they are english or not
#the count keeps track of how many non english characters we find
def non_english_character(string):
    count = 0
    for character in string:
        if ord(character) > 127:
            count += 1
            if count == 3:
                return False
    return True

In [56]:
english_android_clean = [] #empty list to store only the rows for english apps
for row in android_unique:
    name = row[0]
    english = non_english_character(name)
    if english == True:
        english_android_clean.append(row)

len(english_android_clean)   

9597

In [57]:
english_apple_clean = []
for row in apple_store_data:
    name = row[1]
    english = non_english_character(name)
    if english == True:
        english_apple_clean.append(row)
len(english_apple_clean)

6155

## 5)

The last step is to remove all apps that cost money to install, as we are only interested in apps that are free to download and play but consist of in-game ads for revenue.  Let's filter out free apps from our datasets.

In [58]:
ios_free_clean = []
ios_notfree_clean = []
#checks the price column and filters out paid apps
for row in english_apple_clean:
    price = float(row[4])
    if price == 0.0:
        ios_free_clean.append(row)
    else:
        ios_notfree_clean.append(row)
len(ios_free_clean)

3203

In [59]:
android_free_clean = []
android_notfree_clean = []
for row in english_android_clean:
    price = row[6]
    if price == 'Free':
        android_free_clean.append(row)
    else:
        android_notfree_clean.append(row)
len(android_free_clean)

8847

## Data Analysis

Now that our datasets are clean we can begin the analysis.  The first step will be to see what are the most used genres for each dataset, this will help build a profile for the most profitable apps on both the Google Play Store and App Store.  This will involve using the genre columns and number of reviews and average reviews from both datasets.

In [60]:
#this generates a frequency table where the key is the genre and the value is the percentage
#of total apps that are of a particular genre
def freq_table(dataset, index):
    frequency = {}
    total = 0
    for row in dataset:
        genre = row[index]
        total += 1
        if genre in frequency:
            frequency[genre] += 1
        else:
            frequency[genre] = 1
    for genre in frequency:
        frequency[genre] = (frequency[genre]/total)*100
    return frequency

In [63]:
#this uses the frequency table to display the percentage of apps in a particular genre
def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True) #sorts out list from highest percentage to lowest
    for entry in table_sorted:
        print(entry[1], ':', f'{entry[0]:.2f}''%')

In [64]:
apple_genre_freq = display_table(ios_free_clean, 11)


Games : 58.26%
Entertainment : 7.84%
Photo & Video : 5.00%
Education : 3.68%
Social Networking : 3.31%
Shopping : 2.59%
Utilities : 2.47%
Sports : 2.15%
Music : 2.06%
Health & Fitness : 2.03%
Productivity : 1.75%
Lifestyle : 1.56%
News : 1.34%
Travel : 1.25%
Finance : 1.09%
Weather : 0.87%
Food & Drink : 0.81%
Reference : 0.53%
Business : 0.53%
Book : 0.37%
Navigation : 0.19%
Medical : 0.19%
Catalogs : 0.12%


With 58.26% of the downloads, games is by far the most popular app genre.  Also of the top 5, 3 are for entertainment purposes, indicating that entertainment apps get the most usage, and practical purposes such as  health and fitness are less.  Now this is true only for free apps for english users, and doesn't necessarily reflect the entire dataset as a whole.  This also does not reflect the number of users of each app, merely the number of apps in each genre on the Apple App Store.

In [65]:
google_genre_freq = display_table(android_free_clean, 9)

Tools : 8.44%
Entertainment : 6.08%
Education : 5.36%
Business : 4.60%
Productivity : 3.90%
Lifestyle : 3.88%
Finance : 3.71%
Medical : 3.54%
Sports : 3.46%
Personalization : 3.32%
Communication : 3.23%
Action : 3.10%
Health & Fitness : 3.09%
Photography : 2.95%
News & Magazines : 2.80%
Social : 2.67%
Travel & Local : 2.33%
Shopping : 2.25%
Books & Reference : 2.14%
Simulation : 2.05%
Dating : 1.87%
Arcade : 1.84%
Video Players & Editors : 1.77%
Casual : 1.76%
Maps & Navigation : 1.39%
Food & Drink : 1.24%
Puzzle : 1.13%
Racing : 0.99%
Role Playing : 0.94%
Libraries & Demo : 0.94%
Auto & Vehicles : 0.93%
Strategy : 0.90%
House & Home : 0.80%
Weather : 0.79%
Events : 0.71%
Adventure : 0.67%
Comics : 0.60%
Beauty : 0.60%
Art & Design : 0.60%
Parenting : 0.50%
Card : 0.45%
Trivia : 0.42%
Casino : 0.42%
Educational;Education : 0.40%
Board : 0.38%
Educational : 0.37%
Education;Education : 0.34%
Word : 0.26%
Casual;Pretend Play : 0.24%
Music : 0.20%
Racing;Action & Adventure : 0.17%
Puzzle;B

In [66]:
google_category_freq = display_table(android_free_clean, 1)

FAMILY : 18.93%
GAME : 9.70%
TOOLS : 8.45%
BUSINESS : 4.60%
PRODUCTIVITY : 3.90%
LIFESTYLE : 3.89%
FINANCE : 3.71%
MEDICAL : 3.54%
SPORTS : 3.39%
PERSONALIZATION : 3.32%
COMMUNICATION : 3.23%
HEALTH_AND_FITNESS : 3.09%
PHOTOGRAPHY : 2.95%
NEWS_AND_MAGAZINES : 2.80%
SOCIAL : 2.67%
TRAVEL_AND_LOCAL : 2.34%
SHOPPING : 2.25%
BOOKS_AND_REFERENCE : 2.14%
DATING : 1.87%
VIDEO_PLAYERS : 1.80%
MAPS_AND_NAVIGATION : 1.39%
FOOD_AND_DRINK : 1.24%
EDUCATION : 1.16%
ENTERTAINMENT : 0.96%
LIBRARIES_AND_DEMO : 0.94%
AUTO_AND_VEHICLES : 0.93%
HOUSE_AND_HOME : 0.80%
WEATHER : 0.79%
EVENTS : 0.71%
PARENTING : 0.66%
ART_AND_DESIGN : 0.64%
COMICS : 0.61%
BEAUTY : 0.60%


The two frequency tables for the Play store however indicatate that there are more practical apps, plus family oriented apps.  But entertainment still occupies a large portion of the apps.  While it is too early to confidently build an app profile, entertainment still seems to be the safest bet.

In [68]:
ios_genre = freq_table(ios_free_clean, 11)
ios_genre_list = []

for genre in ios_genre:
    total = 0
    len_genre = 0
    for row in ios_free_clean:
        genre_app = row[11]
        if genre_app == genre:
            total += float(row[5])
            len_genre += 1
    ave_n_ratings = total/len_genre
    ios_genre_list.append([ave_n_ratings, genre])
    
for row in sorted(ios_genre_list, reverse = True):
    print(row[1], ':', f'{row[0]:.3f}''%')

Navigation : 86090.333%
Reference : 79350.471%
Social Networking : 71548.349%
Music : 57326.530%
Weather : 52279.893%
Book : 46384.917%
Food & Drink : 33333.923%
Finance : 32367.029%
Photo & Video : 28441.544%
Travel : 28243.800%
Shopping : 27230.735%
Health & Fitness : 23298.015%
Sports : 23008.899%
Games : 22886.367%
News : 21248.023%
Productivity : 21028.411%
Utilities : 19156.494%
Lifestyle : 16815.480%
Entertainment : 14195.359%
Business : 7491.118%
Education : 7003.983%
Catalogs : 4004.000%
Medical : 612.000%


Despite games being the most common genre of app in the App store, navigation, reference and Social Networking top the list in average number of reviews.  Now we do not have a column for number of downloads so number of reviews is the closest we can get to see how popular genres of apps are.  Now our goal is to create an app profile that generates revenue from in app ads.  Ads in a navigation app are more likely to be ignored as people are usually in transit while using them and do not have a time to check out different products.  Social networking and reference apps are next most polular in terms of number of reviews, but just because there are a lot of reviews does not mean that they are good.  Let's now look at the highest rated genres.

In [69]:
apple_genre = freq_table(ios_free_clean, 11)

apple_genre_list = []

for genre in apple_genre:
    total = 0
    len_genre = 0
    for row in ios_free_clean:
        genre_app = row[11]
        if genre_app == genre:
            total += float(row[7])
            len_genre += 1
    ave_ratings = total/len_genre
    apple_genre_list.append([ave_ratings, genre])
    
for row in sorted(apple_genre_list, reverse = True):
    print(row[1], ':', f'{row[0]:.3f}')

Catalogs : 4.125
Games : 4.052
Productivity : 4.000
Shopping : 3.976
Business : 3.971
Music : 3.947
Photo & Video : 3.903
Reference : 3.882
Navigation : 3.833
Health & Fitness : 3.769
Education : 3.636
Food & Drink : 3.635
Social Networking : 3.594
Book : 3.583
Utilities : 3.570
Entertainment : 3.548
Travel : 3.487
Weather : 3.482
Lifestyle : 3.480
Finance : 3.471
News : 3.244
Sports : 3.065
Medical : 3.000


Now let's compare the number of reviews and average reviews for the top 5 genres.

Number of reviews: 
    1. Navigation
    2. Reference
    3. Social Networking
    4. Music
    5. Weather   
Average rating:     
    1. Catalogs
    2. Games
    3. Productivity
    4. Shopping
    5. Business

Due to the sheer volume of games in the app store, creating a new game runs the risk of having that game get lost in the plethora of choices in the app store.  Looking at the highest number of reviews and the highest ratings speaks to a new social media platform specifically geared towards shopping and business.  Let's look at the google play store installs for each genre and see if this idea would play well in the paly store as well!

In [70]:
google_category = freq_table(android_free_clean, 1)

ios_cat_list = []

for category in google_category:
    total = 0
    len_category = 0
    for row in android_free_clean:
        category_app = row[1]
        if category_app == category:
            n_installs = row[5]
            n_installs = n_installs.replace(',', '')
            n_installs = n_installs.replace('+', '')
            n_installs = float(n_installs)
            total += n_installs
            len_category += 1
    ave_n_installs = total/len_category
    ios_cat_list.append([ave_n_installs, category])
    
for row in sorted(ios_cat_list, reverse = True):
    print(row[1], ':', f'{row[0]:.3f}')

COMMUNICATION : 38590581.087
VIDEO_PLAYERS : 24727872.453
SOCIAL : 23253652.127
PHOTOGRAPHY : 17840110.402
PRODUCTIVITY : 16787331.345
GAME : 15544014.510
TRAVEL_AND_LOCAL : 13984077.710
ENTERTAINMENT : 11640705.882
TOOLS : 10830251.971
NEWS_AND_MAGAZINES : 9549178.468
BOOKS_AND_REFERENCE : 8814199.788
SHOPPING : 7036877.312
PERSONALIZATION : 5201482.612
WEATHER : 5145550.286
HEALTH_AND_FITNESS : 4188821.985
MAPS_AND_NAVIGATION : 4049274.634
FAMILY : 3697848.173
SPORTS : 3650602.277
ART_AND_DESIGN : 1986335.088
FOOD_AND_DRINK : 1924897.736
EDUCATION : 1833495.146
BUSINESS : 1712290.147
LIFESTYLE : 1446158.224
FINANCE : 1387692.476
HOUSE_AND_HOME : 1360598.042
DATING : 854028.830
COMICS : 832613.889
AUTO_AND_VEHICLES : 647317.817
LIBRARIES_AND_DEMO : 638503.735
PARENTING : 542603.621
BEAUTY : 513151.887
EVENTS : 253542.222
MEDICAL : 120550.620


## Building an App Profile

The genres with the most number of installs include games, social and photography.  Combining this with the results of the app store analysis give us a clear idea of an app profile.  What we are proposing for an app profile is a social media app that businesses partner with to provide users with discounts.  This app will encourage users to travel the city to get steps and find landmarks.  This app will track location and send the users ads with mystery businesses that will require the user to open to see.  In these ads will be small discounts to local stores that can only be unlocked by reaching a certain number of steps or travelling to a certain landmark nearby and taking a picture to upload to the app. Following businesses and uploading pics of users in front of those businesses will also unlock discounts for the user.